File: LC543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = 0

    def maxDepth(self, root):
        if root == None:
            return 0
        
        return 1 + max(self.maxDepth(root.left),self.maxDepth(root.right))
    
    def pre(self, root):
        if root == None:
            return
        
        count = self.maxDepth(root.left) + self.maxDepth(root.right)
        if self.ans < count:
            self.ans = count
        
        self.pre(root.left)
        self.pre(root.right)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        if not root.left and not root.right:
            return 0

        self.pre(root)

        return self.ans

File: LC543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == NULL)
            return 0;
        
        if(!root->left && !root->right)
            return 0;
        
        TreeNode *t = root;
        
        int ans = 0;

        pre(root, ans);

        return ans;

    }
private:
    void pre(TreeNode *root, int &ans){

        if(root == NULL)
            return;

        int c = maxDepth(root->left) + maxDepth(root->right);
        if(ans < c)
            ans = c;
        pre(root->left, ans);
        pre(root->right, ans);
    }

    int maxDepth(TreeNode* root){
        if(root == NULL)
            return 0;
        
        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
};
